src/FreetypeRenderer.cc

```diff
@@ -97,49 +95,105 @@ int FreetypeRenderer::outline_cubic_to_func(const FT_Vector *c1, const FT_Vector
+            params.halign);
+        x_offset = 0;
+    }
+
+    if (params.valign == "top") {
+        y_offset = -ascent;
+    } else if (params.valign == "center") {
+        double height = ascent - descent;
+        y_offset = -height/2 - descent;
+    } else if (params.valign == "bottom") {
+        y_offset = -descent;
+    } else if (params.valign == "baseline" || params.valign == "default") {
+        y_offset = 0;
+    } else {
+        LOG(message_group::Warning, params.loc, params.documentPath,
+            "Unknown value for the valign parameter"
+            " (use \"baseline\", \"bottom\", \"top\" or \"center\"): '%1$s'",
+            params.valign);
+        y_offset = 0;
+    }
 }
 
-double FreetypeRenderer::calc_y_offset(std::string valign, double ascend, double descend) const
+// Calculate offsets for vertical text.
+void FreetypeRenderer::ShapeResults::calc_offsets_vert(
+    const FreetypeRenderer::Params &params)
 {
-	if (valign == "top") {
-		return -ascend;
-	} else if (valign == "center") {
-		return descend / 2.0 - ascend / 2.0;
-	} else if (valign == "bottom") {
-		return descend;
-	} else {
-		if (valign != "baseline") {
-			LOG(message_group::None,Location::NONE,"","Unknown value for the valign parameter (use \"baseline\", \"bottom\", \"top\" or \"center\"): '%1$s'",valign);
-		}
-		return 0;
-	}
+    if (params.halign == "right") {
+        x_offset = -right;
+    } else if (params.halign == "left") {
+        x_offset = -left;
+    } else if (params.halign == "center" || params.halign == "default") {
+        x_offset = 0;
+    } else {
+        LOG(message_group::Warning, params.loc, params.documentPath,
+            "Unknown value for the halign parameter"
+            " (use \"left\", \"right\" or \"center\"): '%1$s'",
+            params.halign);
+        x_offset = 0;
+    }
+
+    if (params.valign == "baseline") {
+        LOG(message_group::Warning,params.loc,params.documentPath,
+                    "Don't use valign=\"baseline\" with vertical layouts",
+                    params.valign);
+        y_offset = 0;
+    } else if (params.valign == "center") {
+        y_offset = -advance_y / 2.0;
+    } else if (params.valign == "bottom") {
+        y_offset = -advance_y;
+    } else if (params.valign == "top" || params.valign == "default") {
+        // Note that in vertical mode HarfBuzz sets the glyphs
+        // below their origins, so this results in the entire string
+        // being placed below the origin.
+        y_offset = 0;
+    } else {
+        LOG(message_group::Warning, params.loc, params.documentPath,
+            "Unknown value for the valign parameter"
+            " (use \"baseline\", \"bottom\", \"top\" or \"center\"): '%1$s'",
+            params.valign);
+    }
 }
 
-hb_direction_t FreetypeRenderer::get_direction(const FreetypeRenderer::Params &params, const hb_script_t script) const
+hb_direction_t FreetypeRenderer::Params::detect_direction(const hb_script_t script) const
```

Comment:
I was bothered that `get_direction` and `get_script` were not simple getters; they were not even symmetric with the type accepted by `set_direction` and `set_script`.

I renamed them to `detect_xxx` to make the distinction a little clearer, but I'm still not sure that the structure is what it should be.  It's tempting to have the string value be *only* the value passed in, and have a separate `hb_direction_t` / `hb_script_t` value that's the result of heuristics, defaulting, et cetera.  In that model, `detect_xxx` would probably move over into `ShapeResults` and would just set `hb_xxx` values there.  That model would avoid the mildly silly behavior of taking a string, converting (and defaulting, applying heuristics, et cetera) it into a HB enum, converting that enum back into a string, and then finally converting the string into a HB enum again.  The downside would be that the final value wouldn't be directly printable, but the HB enum could be converted if desired.  (But:  Maybe it would also affect what shows up in CSG dumps, making them show the input value rather than the final value.)